mettle/src/stdapi/sys/memory.c

```diff
@@ -0,0 +1,503 @@
+#include <stdio.h>
+#include <regex.h>
+#include <fnmatch.h>
+#include <fnmatch.h>
+#include <ctype.h>
+
+#include "channel.h"
+#include "log.h"
+#include "tlv.h"
+#include "command_ids.h"
+#include "memory.h"
+
+#define NEEDLES_MAX 5
+#define MAX_ADDR_DATA 500
+#define MATCH_LEN_MAX 250
+#define MATCH_PER_NEEDLE_MAX 5
+
+#define MAX_STRINGS 350
+#define MIN_SEARCH_LEN 5
+
+struct tlv_packet *mem_read(struct tlv_handler_ctx *ctx)
+{
+	uint64_t pid;
+	uint32_t size;
+	uint64_t start_addr;
+	struct tlv_packet *p = NULL;
+
+	if(tlv_packet_get_u64(ctx->req, TLV_TYPE_HANDLE, &pid) == -1)
+	{
+		log_debug("Pid was not retrieved.\n");
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	if(tlv_packet_get_u64(ctx->req, TLV_TYPE_BASE_ADDRESS, &start_addr) == -1)
+	{
+		log_debug("Failed to retrieve the start address.\n");
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	if(tlv_packet_get_u32(ctx->req, TLV_TYPE_LENGTH, &size) == -1)
+	{
+		log_debug("Failed to retrieve size to search.\n");
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	if(size <= 0)
+	{
+		size = MATCH_LEN_MAX;
+	}
+
+	int path_len = strlen("/proc/") + 8 + strlen("/mem") + 1;
+
+	char *read_str = NULL;
+	char mem_path[path_len];
+	snprintf(mem_path, path_len, "/proc/%zu/mem", (size_t)pid);
+	log_debug("Path to read from: %s\n", mem_path);
+
+	FILE *fp = fopen(mem_path, "r");
+	if(fp == NULL)
+	{
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	unsigned char buf[size];
+	fseeko(fp, start_addr, SEEK_SET);
+	fread(buf, 1, size, fp);
+	fclose(fp);
+
+	p = tlv_packet_response_result(ctx, TLV_RESULT_SUCCESS);
+	p = tlv_packet_add_raw(p, TLV_TYPE_PROCESS_MEMORY, buf, size);
+
+	return p;
+}
+
+struct tlv_packet *mem_search(struct tlv_handler_ctx *ctx)
+{
+	uint32_t pid;
+	uint32_t match_len;
+	uint32_t min_search_len;
+	struct tlv_packet *p = NULL;
+
+	if(tlv_packet_get_u32(ctx->req, TLV_TYPE_PID, &pid) == -1)
+	{
+		log_debug("Failed to retrieve PID\n");
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	if(tlv_packet_get_u32(ctx->req, TLV_TYPE_MEMORY_SEARCH_MATCH_LEN, &match_len) == -1)
+	{
+		log_debug("Failed to retrieve match length\n");
+		return tlv_packet_response_result(ctx, TLV_RESULT_FAILURE);
+	}
+
+	if(tlv_packet_get_u32(ctx->req, TLV_TYPE_UINT, &min_search_len) == -1)
+	{
+		log_debug("Using default minimum for search\n");
+		min_search_len = MIN_SEARCH_LEN;
+	}
+
+	if(match_len > MATCH_LEN_MAX)
+	{
+		match_len = MATCH_LEN_MAX;
+	}
+
+	log_debug("PID: %zu\n", (size_t) pid);
+	struct addr_range *ranges = parse_maps_file(pid);
```

Comment:
If this isn't being freed anywhere, we should probably free it within this function.